test/cli/tarantool_leak.c

```diff
@@ -0,0 +1,52 @@
+#include "test.h"
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <tarantool/tarantool.h>
+#include <tarantool/tnt_net.h>
+#include <tarantool/tnt_opt.h>
+#include <msgpuck.h>
+
+#include "common.h"
+
+const char template[] = "[{\"m\": {\"m\": [], \"m\": [], \"m\": [], \"m\": [], \"m\": {\"m\": [\"a\", \"a\"]}}}]";
+
+static int
+test_leak(void)
+{
+        plan(1);
+
+        char data[sizeof(template)] = { 0 };
+        struct tnt_stream * stream = tnt_object(NULL);
+        tnt_object_add_array(stream, 1);
+        tnt_object_add_map(stream, 1);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_map(stream, 5);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_array(stream, 0);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_array(stream, 0);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_array(stream, 0);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_array(stream, 0);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_map(stream, 1);
+        tnt_object_add_str(stream, "m", 1);
+        tnt_object_add_array(stream, 2);
+        tnt_object_add_str(stream, "a", 1);
+        tnt_object_add_str(stream, "a", 1);
+        mp_snprint(data, sizeof(data), TNT_SBUF_DATA(stream));
+        ok(!strcmp(data, template), "Invalid msgpuck");
+        tnt_stream_free(stream);
+
+        return check_plan();
+}
+
+int main()
+{
+        test_leak();
+        return 0;
+}
```

Comment:
Writing a test that always passes is bad. IMO better mock the allocation function. The allocation function used by the connector is already virtual so you could replace it in the test to account memory.

Reply:
Fixed, implement function which saves allocated pointers in std:set. In test i check that set is empty after resource release.